packages/auth/src/core/auth/auth_impl.ts

```diff
@@ -457,10 +453,102 @@ export class AuthImpl implements AuthInternal, _FirebaseService {
       );
     }
 
-    const passwordPolicy: PasswordPolicy = {
+    let validFlag = true;
```

Comment:
I will move them into helper functions within the auth class, since these methods will likely not be used outside of this class. I feel it's fine to have two helper methods for now, one for checking the length options and another for checking the character options.

Reply:
Actually, since `PasswordValidationStatus` is read only, it doesn't look like we can assign each property as we check them within the helper methods. I don't think we should create a separate method for each requirement since each would have to iterate over the password (as you mentioned). I could create a `PasswordPolicyBuilder` class to build the status as we check, but I'm not sure if this would complicate the design more than we want.

Reply:
How about adding a list of validation helper methods on PasswordPolicyInternal?

Reply:
We could do that and have `PasswordPolicyInternalImpl` class that implements the interface and defines the helper methods. This still has the same issue where we would either have to have a helper method for each requirement or store flags internally within the class and have a final builder method for the status since the fields of `PasswordValidationStatus` are readonly. I think having a separate method for each requirement would make this implementation cleaner, but that would mean we have to iterate over the password multiple times to check each of the character requirements. This may not be that big of a concern since the passwords will be at most 4096 characters.

Reply:
Based on our conversation offline, I'm going to create an internal version of both `PasswordPolicy` and `PasswordValidationStatus`. The internal version of the validation status will not be readonly, and the internal policy will have helper methods for checking password requirements. I'll split this PR into multiple since this includes a lot of changes.